DESIGN.md

```diff
@@ -0,0 +1,89 @@
+# Design Overview
+
+## Primary Tools
+
+**Frontend**: `Javascript`/`React`/`Webpack`/Airbnb-styleguide -- a standard frontend stack in keeping with Teleport's typical stack. For the sake of expediency I'll build for the latest stable Chrome release and will ignore mobile optimization.
+
+**Backend**: `Go 1.15` standard library plus external dependencies (discussed below)
+
+## Router/Dispatcher
+
+The router/dispatcher will use the popular [gorilla/mux](https://github.com/gorilla/mux) package rather than the standard `http.ServeMux`. The primary reason for this choice is that `gorilla` provides a cousin package [gorilla/websocket](https://github.com/gorilla/websocket) for handling websockets. Technically `gorilla/websocket` should be compatible with `go`'s standard `http` package, however using the two `gorilla` packages together ensures superior online support and avoids any possible idiosyncratic incompatibilities. Additionally `gorilla` seemingly has a reputation for better developer ergonomics than the standard package.
+
+Note: My plan is to initially implement the system with client-side polling (the level 2 project), and then time/energy permitting attempt to refactor the system to hit the level 3/4 project which would make use of websockets.
+
+## Browser-client Security
+
+Browser-clients will use a username/password to authenticate and then be granted an access token which will serve as a session ID and used to authorize subsequent requests (sent in the Authorization header in Bearer Token format). The server will be built to only accept SSL encrypted connections which will prevent (all but the most sophisticated) MITM attackers from stealing the token from the header and/or other information sent over the wire.
```

Comment:
I would also add HTML headers to APIs such as:

```
X-XSS-Protection
X-Frame-Options
Strict-Transport-Security
```

Reply:
Good catch. I'll add

```
X-XSS-Protection: 1; mode=block
X-Frame-Options: deny
Strict-Transport-Security: max-age= 6307200; includeSubDomains
Content-Security-Policy: default-src 'self'
X-Permitted-Cross-Domain-Policies: none
```
(Taken from [here](https://wiki.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Headers))